### PR TITLE
Handle xdg popups and subsurfaces

### PR DESCRIPTION
**Pull request by tiosgz**
I started this, because it happened quite often that menues were off-screen. Damage tracking (in its current state) completely ignores them, so that’s one more issue fixed here. One last thing was rendering them above (almost) everything, but I think that can wait for another PR (I’m not sure how soon or late it would be done).

A problem that’s still present is that subsurfaces are moved into the output only for the left (0,0) output, but it’s the same in Sway so I guess I can’t do anything about it.

I have left a few questions around in the code (`grep -rF 'Q:'`). Also, I don’t know how much of the code is unused, but some of it may be useful later (though I can remove it of course).

// TODO credit Sway for about half of the code

### CODE REVIEW

**Comment by tiosgz** [comment]
Damn, GH detected even `part-fix` as a keyword. Well, just letting you know that if you want to keep #38 open until it’s completely fixed, you should unlink it.

**Comment by tiosgz** [comment]
TODO: handle moving everything -> unconstrain / send_enter again.

**Review by buffet — CHANGES_REQUESTED** [review]
If we set child->map when the owner/root view gets mapped/unmapped, we incorrectly restore that when it's mapped again.

Same with sub-subsurfaces, if those get unmapped we don't unmap the child, it would make more sense if we crawl up the tree and check if they're all mapped, similar to how sway does it.

**Comment by buffet** [comment]
Thanks for all this work!

**Comment by buffet** [comment]
Oh, also how does this behave when the view is moved (from lua)?

And is there any specific app you're using to test this?

File: include/desktop/view.h [hunk 1]
```diff
@@ -88,6 +90,45 @@ struct kiwmi_view_impl {
...
+    bool mapped;
+
+    struct wl_listener commit;
+    struct wl_listener map;
+    struct wl_listener unmap;
+    struct wl_listener view_unmap;
+    struct wl_listener new_popup;
+    struct wl_listener new_subsurface;
+    // Q: have got a better name?
+    struct wl_listener owner_destroy;   // the union'ed object destroy
```
[the remark this answers is not in the document]
Comment (tiosgz) [reply include/desktop/view.h:123]: It’s a destroy listener for the wlr_xdg_popup or wlr_subsurface.

(Sway listens for destroy at both the wlr_surface and the popup/subsurface, so I thought it’s necessary for both. I don’t know if it really is though.)

File: kiwmi/desktop/view.c [hunk 2]
```diff
@@ -259,3 +284,191 @@ view_create(
...
+    struct kiwmi_output *output;
+    wl_list_for_each (output, &child->view->desktop->outputs, link) {
+        output_damage(output);
+    }
+}
+
+void
+view_child_destroy(struct kiwmi_view_child *child)
+{
+    bool mapped = child->mapped && child->view->mapped && !child->view->hidden;
```
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/view.c:300]: It’s only used for damage, so `hidden` is important as well. Maybe `visible` (or `rendered`, coz it isn’t visible when it’s behind another surface) is a better name for the local variable.

File: kiwmi/desktop/view.c [hunk 3]
```diff
@@ -259,3 +284,191 @@ view_create(
...
+    child->commit.notify = view_child_commit_notify;
+    wl_signal_add(&wlr_surface->events.commit, &child->commit);
+
+    child->surface_destroy.notify = view_child_surface_destroy_notify;
+    wl_signal_add(&wlr_surface->events.destroy, &child->surface_destroy);
+
+    child->new_subsurface.notify = view_child_new_subsurface_notify;
+    wl_signal_add(&wlr_surface->events.new_subsurface, &child->new_subsurface);
+
+    child->view_unmap.notify = view_child_view_unmap_notify;
```
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/view.c:427]: I was copying this from Sway. It only listens to `unmap` on the view. But looking at https://github.com/swaywm/sway/commit/017a7c4da169ace601f282c87a9aaabc138a34d3, there seems to be no reason not to handle `map` as well.

File: kiwmi/desktop/view.c [hunk 4]
```diff
@@ -259,3 +284,191 @@ view_create(
...
+    wl_signal_add(&view->events.unmap, &child->view_unmap);
+
+    // Let the caller add the listeners to the respective events
+    child->map.notify = view_child_map_notify;
+    wl_list_init(&child->map.link);
+
+    child->unmap.notify = view_child_unmap_notify;
+    wl_list_init(&child->unmap.link);
+
+    wl_list_init(&child->new_popup.link);
```
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/view.c:437]: `new_popup` is only used in `xdg_popup`, so I’d say that at least it should (though there’s also the `calloc` 🤷)

I’d rather keep it for `owner_destroy` as well, since a new `view_child` could potentially not have it.

File: kiwmi/desktop/view.c [hunk 5]
```diff
@@ -259,3 +284,191 @@ view_create(
...
+
+struct kiwmi_view_child *
+view_child_subsurface_create(
+    struct kiwmi_view_child *parent,
+    struct kiwmi_view *view,
+    struct wlr_subsurface *subsurface)
+{
+    struct kiwmi_view_child *child = view_child_create(
+        view, subsurface->surface, KIWMI_VIEW_CHILD_SUBSURFACE, NULL);
+    if (child == NULL) {
```
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/view.c:453]: Sure! Was copying from Sway…

File: kiwmi/desktop/view.c [hunk 6]
```diff
@@ -259,3 +284,191 @@ view_create(
...
+        wl_list_insert(&parent->children, &child->link);
+    }
+
+    wl_signal_add(&subsurface->events.map, &child->map);
+    wl_signal_add(&subsurface->events.unmap, &child->unmap);
+
+    child->owner_destroy.notify = view_child_subsurface_owner_destroy_notify;
+    wl_signal_add(&subsurface->events.destroy, &child->owner_destroy);
+
+    child->mapped = true;
```
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/view.c:470]: This is what Sway does. I’ll check the specs.

File: include/desktop/view.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (buffet) [reply include/desktop/view.h:123]: Ah, hm I see. I'll try to think of something, though this smells like "there is no good name" already.

File: kiwmi/desktop/view.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (buffet) [reply kiwmi/desktop/view.c:300]: Ahh, I see, yeah that name seems better then.

How does this behave when the view is mapped but the child's parent isn't? (I think this ties together with my other comment)

File: kiwmi/desktop/view.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (buffet) [reply kiwmi/desktop/view.c:427]: Ok epic. I'd say don't handle if we don't need to.

File: kiwmi/desktop/view.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (buffet) [reply kiwmi/desktop/view.c:437]: Ah, I see. Yeah with the remove stuff better keep it. Good catch!

**Comment by tiosgz** [comment]
> If we set child->map when the owner/root view gets mapped/unmapped, we incorrectly restore that when it's mapped again.

(see the other thread on this)

> Same with sub-subsurfaces, if those get unmapped we don't unmap the child, it would make more sense if we crawl up the tree and check if they're all mapped, similar to how sway does it.

I’m not sure I understand you here. But if it’s just for consistency with view unmapping, the message of the Sway commit I linked above can maybe explain why it isn’t needed.

> Thanks for all this work!

I must admit (again and again) that I couldn’t have done it without Sway, because I wouldn’t know what and where to do. And because I’m miserably bad at >100loc ‘atomic’ changes :P

> Oh, also how does this behave when the view is moved (from lua)?

It doesn’t do anything AFAICT. See the TODO above ;)

> And is there any specific app you're using to test this?

I don’t know of one that would be guaranteed to behave correctly, but I haven’t done any research (one more TODO, yay). For now Sublime Merge (coz its button titles were what I saw misbehave, although it’s partly due to SM too) and Vivaldi (that was useful just to check the damage tracking, coz I can show the menu without anything else changing).

PS: I’ll keep adding commits until it’s ready for merge and stash them at the last moment. It’s a bit more convenient for me, except I have to make sure you know this so you don’t merge an ugly commit history.

File: kiwmi/desktop/view.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/view.c:300]: The actual rendering hasn’t been affected at all in this PR.

File: kiwmi/desktop/view.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/view.c:427]: I wonder if it’s just that views never unmap other than before destroying or if it’s completely correct.

(Marking as unresolved only so that you see this, feel free to re-resolve.)

File: kiwmi/desktop/view.c [hunk 7]
```diff
@@ -229,6 +229,31 @@ view_resize(struct kiwmi_view *view, uint32_t edges)
     view_begin_interactive(view, KIWMI_CURSOR_RESIZE, edges);
 }
 
+void
+// Q: how to call it? In sway it's view_init_subsurfaces and view_child_init_subsurfaces.
+// Or keep it split into two almost identical functions?
+view_init_subsurfaces(struct kiwmi_view_child *child, struct kiwmi_view *view)
```
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/view.c:235]: It’s just that it can be confusing w/o looking at the implementation and/or usage.

File: include/desktop/view.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (tiosgz) [reply include/desktop/view.h:122]: Maybe `[surface_]extension`, because that’s what they are?

File: kiwmi/desktop/view.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/view.c:423]: I haven’t found where these (or rather, the code that calls them) are called, but wlroots seem to handle parents for us (so the only thing we have to care about is the invalid pointer).

https://github.com/swaywm/wlroots/blob/46c42e55c6e0b08d9e7db989d6a10f073525e999/types/wlr_surface.c#L1026..L1080

File: kiwmi/desktop/view.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/view.c]: Specs:

> A sub-surface becomes mapped, when a non-NULL wl_buffer is applied and the parent surface is mapped. The order of which one happens first is irrelevant. A sub-surface is hidden if the parent becomes hidden, or if a NULL wl_buffer is applied. These rules apply recursively through the tree of surfaces.

It seems to be handled by wlroots, and when I removed it, Sublime Merge tooltips still work (I probably really need a well-behaved but edge-case client 😉).

[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/view.c]: I’m dumb. Of course during my ‘test’ nothing was broken, because `child->mapped` is only read by the destructor, and damaging twice (on creation and destruction) should have the same effect as not damaging (in the edge case when the subsurface isn’t really used).

**Comment by tiosgz** [comment]
Added a TODO at the top. The draft PR state is just for consistency with the checklist.

File: include/desktop/view.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (buffet) [reply include/desktop/view.h:122]: They are the extensions, not the owner though, unless I'm missing something.

**Comment by tiosgz** [comment]
Sorry if it doesn’t make sense. I still do believe that wlroots handle all parent (un)maps, but I can’t find the proof. This is an approximate reflection of my thoughts (only being added to as it’s becoming clearer to me, I didn’t remove anything) before I’ll be sure.

I think it’s all done by wlroots, Sway also only cares about the child’s own map/unmap and the view’s unmap (just because of the use-after-free). Though I’m still stuck with [`subsurface_consider_map`](https://github.com/swaywm/wlroots/blob/46c42e55c6e0b08d9e7db989d6a10f073525e999/types/wlr_surface.c#L1026..L1062) and `subsurface_unmap` – that is, I only see it bubbling through the subsurfaces, but not the initial trigger on view map (only on subsurface commit). (Are all surfaces expected to commit even if they were unmapped by their parent when they want to map again? Wlroots commit on all *synchronized* child surfaces automatically.)

So the question is left to desynchronized subsurfaces only.

---
This is maybe just a link dump for myself, but could be useful

descendant popups are destroyed on unmap
https://github.com/swaywm/wlroots/blob/46c42e55c6e0b08d9e7db989d6a10f073525e999/types/xdg_shell/wlr_xdg_surface.c#L33..L36

TODO: check that the unmap damaging cannot use-after-free too
https://github.com/swaywm/wlroots/blob/46c42e55c6e0b08d9e7db989d6a10f073525e999/types/wlr_surface.c#L1136..L1140

wlroots commit on all *synchronized* child surfaces when their parent commits, but there’s nothing about desynchronized ones
https://github.com/swaywm/wlroots/blob/46c42e55c6e0b08d9e7db989d6a10f073525e999/types/wlr_surface.c#L579..L584

File: include/desktop/view.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (tiosgz) [reply include/desktop/view.h:122]: So if there’s no better name, `extension_destroy` is the way to go?

File: kiwmi/desktop/xdg_shell.c [hunk 8]
```diff
@@ -22,6 +23,123 @@
...
+{
+    if (popup->type != KIWMI_VIEW_CHILD_XDG_POPUP) {
+        wlr_log(WLR_ERROR, "Expected an xdg_popup kiwmi_view_child");
+        return;
+    }
+
+    struct kiwmi_view *view         = popup->view;
+    struct wlr_xdg_popup *wlr_popup = popup->wlr_xdg_popup;
+
+    // Q: are the coord calculations across this function correct? I have
```
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/xdg_shell.c:59]: This is how I see it now, with help from Sway’s [`subsurface_get_root_coords`](https://github.com/swaywm/sway/blob/bb3fd0abc56f39e35dc7f4e86f25da1b4a6efbd7/sway/tree/view.c#L904)

```
,-- ly=0
|   (nothing from the view is displayed here)
| . . vy == -view->geom.y; toplevel surface begin
| .   (e.g. drop-shadow)
| . ,-- vy=0; ly == view->y; "user-visible" view area begin
| . |   (this is the actual window)
| . | ,-- sy=0; vy == surf->y; subsurface begin (sy and surf refer to the subsurface here)
| . | |   (here is the subsurface)
```

which also seems to match our rendering (note that it has `output_ly` negative)

```c
int oy = rdata->output_ly + sy + view->y - view->geom.y;
```

(just writing it here because I realised that one of their commits I wanted to look at introduced `subsurface_get_root_coords` and this was the only reason for rechecking it I could think of)

**Comment by HeavyRain266** [comment]
> Sorry if it doesn’t make sense. I still do believe that wlroots handle all parent (un)maps, but I can’t find the proof.

Nope, you have to map, unmap, commit and uncommit stuff yourself everytime and actually doesn't matter if it's wlroots or libwayland, wlrotos isn't that much high level to do stuff like that for yourself.

> I think it’s all done by wlroots, Sway also only cares about the child’s own map/unmap and the view’s unmap (just because of the use-after-free). Though I’m still stuck with subsurface_consider_map and subsurface_unmap – that is, I only see it bubbling through the subsurfaces, but not the initial trigger on view map (only on subsurface commit). (Are all surfaces expected to commit even if they were unmapped by their parent when they want to map again? Wlroots commit on all synchronized child surfaces automatically.)

Honestly, from my experience Sway isn't that good reference, even if it's made by authors of wlroots. They do a lot of stuff in different way than kiwmi or whatever you would like to do it and without studying whole codebase and trying to tralasnate it to your own desired takes too much time imo, from my expiernce I was always looking into something like hikari, taiwins and river (requires some Zig knowledge) and other newer and relativery smaller compositors. (I'm using libwayland anyways so for me everything is strange and too high level, I might wrong so don't mind that)

I'd tried implementing 2-3 protocols based on sway and it was pretty hard to get into, it was idle inhibitor (for idle management), xdg-activation and something else also ralatively small and couldn't even get it to work by looking at sway's code. So will read more kiwmi code and do everything myself.

**Comment by tiosgz** [comment]
> Nope, you have to map, unmap, commit and uncommit stuff yourself everytime and actually doesn't matter if it's wlroots or libwayland, wlrotos isn't that much high level to do stuff like that for yourself.

The last [link](https://github.com/swaywm/wlroots/blob/46c42e55c6e0b08d9e7db989d6a10f073525e999/types/wlr_surface.c#L579..L584) shows that I only need about mapping desynchronized subsurfaces IIUC. If you can deduct something different from it, let me know 😉

> Honestly, from my experience Sway isn't that good reference, even if it's made by authors of wlroots. They do a lot of stuff in different way than kiwmi or whatever you would like to do it and without studying whole codebase and trying to tralasnate it to your own desires takes too much time imo, from my expiernce I was always looking into something like hikari, taiwins and river (requires some Zig knowledge) and other newer and relativery smaller compositors.

This seemed relatively easy to pick from Sway. Anyway, thanks for the suggestions, I’ll have a look at them next time I’m lost 👍

File: include/desktop/view.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (buffet) [reply include/desktop/view.h:122]: Yup, I think so.

File: kiwmi/desktop/view.c [hunk 9]
```diff
@@ -259,3 +299,229 @@ view_create(
...
+    child->view        = view;
+    child->wlr_surface = wlr_surface;
+    child->mapped      = false;
+    // Q: does it make sense to cache it all the time just to avoid the invalid
+    // child->view pointer? Any better idea?
+    // We could destroy all children once they lose connection to their view
+    // (i.e. the view or their parent is destroyed), because if the surfaces are
+    // reused, we create new kiwmi_view_child'ren anyway.
+    child->parent_mapped =
+        parent ? parent->mapped && parent->parent_mapped : view->mapped;
```
Comment (tiosgz) [inline kiwmi/desktop/view.c:457]: I was just lazy to think it through at the time of writing. That said, comments are welcome, but maybe I can sort it out myself.

**Comment by tiosgz** [comment]
The behaviour after 7476079c51c91bb8f720e3e932e71e1d8a823362 is different from most compositors I’ve tried (sway, hikari, river, taiwins; but wayfire behaves the same; I have yet to try weston; unfortunately I have no other compositors currently). But I believe it is correct, so it should be okay? (Then it would be time for me to fix those compositors.)

I currently use [this](https://gist.github.com/tiosgz/bd39364486ce57fccc1728a89a362c77) for testing (sorry, the code is not nice to read). Build intructions are inside.
Expected behaviour: 1 sec black + red subsurfaces, 1 sec none, 1 sec black + red, 1 sec black + orange, repeat.

File: kiwmi/desktop/xdg_shell.c [hunk 10]
```diff
@@ -22,6 +23,158 @@
...
+
+    double popup_sx;
+    double popup_sy;
+    wlr_xdg_popup_get_position(wlr_popup, &popup_sx, &popup_sy);
+
+    // TODO: possibly ignore the popup's own position. When moving a view from
+    // the very right of an output to the very left, popups may end up on an
+    // output to the left of the view's output.
+    struct wlr_output *output = wlr_output_layout_output_at(
+        view->desktop->output_layout, view->x + popup_sx, view->y + popup_sy);
```
Comment (tiosgz) [inline kiwmi/desktop/xdg_shell.c:67]: What would you say is better? Normally I would expect ignoring the popup’s position to be okay, but it can have unexpected results when a view spans multiple outputs.

File: kiwmi/desktop/xdg_shell.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/xdg_shell.c]: I’m quite sure now that it is correct here. Feel free to reopen this if you see it differently.

File: kiwmi/desktop/view.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (buffet) [reply kiwmi/desktop/view.c:247]: Still would prefer a comment about that over duplicating code.

File: kiwmi/desktop/view.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (buffet) [reply kiwmi/desktop/view.c]: Should be good then. :+1:

**Comment by tiosgz** [comment]
FIXME: librewolf mostly frozen (not a problem with 9d2c3adb6e52ab05d28a99d2c3e1d60b09458c1b, likely caused by 7476079c51c91bb8f720e3e932e71e1d8a823362)

File: kiwmi/desktop/view.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/view.c]: Well, that isn’t true anymore. I’ve changed the rendering as well. But it still doesn’t use the new struct (yet).

File: kiwmi/desktop/view.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (tiosgz) [reply kiwmi/desktop/view.c:251]: See 168211d58ed83d5c97a93c31deeb966c7c2c2a1c (is the comment clear enough?)

**Comment by HeavyRain266** [comment]
Subsurfaces, popups and xwayland are probably most annoying part of compositor...

**Comment by buffet** [comment]
> The behaviour after [7476079](https://github.com/buffet/kiwmi/commit/7476079c51c91bb8f720e3e932e71e1d8a823362) is different from most compositors I’ve tried (weston, sway, hikari, river, taiwins; but wayfire behaves the same; unfortunately I have no other compositors currently). But I believe it is correct, so it should be okay? (Then it would be time for me to fix those compositors.)
> 
> I currently use [this](https://gist.github.com/tiosgz/bd39364486ce57fccc1728a89a362c77) for testing (sorry, the code is not nice to read). Build intructions are inside.
> Expected behaviour: 1 sec black + red subsurfaces, 1 sec none, 1 sec black + red, 1 sec black + orange, repeat.

How exactly is it different?

**Comment by tiosgz** [comment]
> > The behaviour after [7476079](https://github.com/buffet/kiwmi/commit/7476079c51c91bb8f720e3e932e71e1d8a823362) is different from most compositors I’ve tried (weston, sway, hikari, river, taiwins; but wayfire behaves the same; unfortunately I have no other compositors currently). But I believe it is correct, so it should be okay? (Then it would be time for me to fix those compositors.)
> > I currently use [this](https://gist.github.com/tiosgz/bd39364486ce57fccc1728a89a362c77) for testing (sorry, the code is not nice to read). Build intructions are inside.
> > Expected behaviour: 1 sec black + red subsurfaces, 1 sec none, 1 sec black + red, 1 sec black + orange, repeat.
> 
> How exactly is it different?

The other compositors ignore that the black subsurface (which is the parent of the red/orange one) is unmapped and render the child anyway.
It becomes: 1 sec black + red subsurfaces, 1 sec red, 1 sec black + red, 1 sec black + orange, repeat.

File: kiwmi/desktop/xdg_shell.c (hunk 10 above)
Comment (buffet) [reply kiwmi/desktop/xdg_shell.c:67]: I think not taking the output into account makes sense here.

I'm not sure what's the best approach here is, do you know what the other compositors do?

Comment (buffet) [reply kiwmi/desktop/xdg_shell.c:67]: I think not taking the output into account makes sense here.

I'm not sure what's the best approach here is, do you know what the other compositors do?

File: kiwmi/desktop/view.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (buffet) [reply kiwmi/desktop/view.c:251]: Seems good. :+1:

File: kiwmi/desktop/view.c [hunk 11]
```diff
@@ -369,9 +369,10 @@ static void
 view_child_commit_notify(struct wl_listener *listener, void *UNUSED(data))
 {
     struct kiwmi_view_child *child = wl_container_of(listener, child, commit);
-    if (child->mapped && child->parent_mapped) {
-        view_child_damage(child);
-    }
+    // TODO: when can it map and not affect .mapped and/or .parent_mapped?
```
Comment (buffet) [inline kiwmi/desktop/view.c:376]: I'd say it can't, but well... I'll try to do some testing with this myself.

File: kiwmi/desktop/view.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (buffet) [reply kiwmi/desktop/view.c]: I'm going through everything rn anyways, haha.

File: kiwmi/desktop/view.c (hunk 11 above)
Comment (tiosgz) [reply kiwmi/desktop/view.c:376]: It was just a quick note for myself, I didn’t check where the problem could lie – I was interested in the workaround primarily. It also shows that it may really be a bad idea to cache `parent_mapped`, but until I know the problem I can’t really say.

Anyway, I’ll be glad for any hint.

File: kiwmi/desktop/view.c (hunk 9 above)
Comment (buffet) [reply kiwmi/desktop/view.c:462]: If they get recreated anyways, just destroy them I'd say. Am I missing something?

File: kiwmi/desktop/view.c (hunk 11 above)
Comment (tiosgz) [reply kiwmi/desktop/view.c]: Apparently fixed

File: kiwmi/desktop/view.c (hunk 9 above)
Comment (tiosgz) [reply kiwmi/desktop/view.c]: I was just unsure if I saw everything.

File: kiwmi/desktop/xdg_shell.c (hunk 10 above)
Comment (tiosgz) [reply kiwmi/desktop/xdg_shell.c:67]: Another possibility is to give control to the user, e.g. via `view:unconstrain_rect(lx, ly, w, h)` (or `vx, vy`) defaulting to the view’s rect.

That way this function doesn’t have to be so complicated and the user can always do what they wish.

Comment (tiosgz) [reply kiwmi/desktop/xdg_shell.c:67]: Hmm, that is also not a good idea.

If we use `vx, vy`, it’s rather illogical, because a user is most likely to want to use an output’s rect, which is in `lx, ly`, and they have to update the unconstrain rect every time they move a view.

If we use `lx, ly`, it can become unpredictable. A user is likely to expect the unconstrain rect to be applied automatically when the view moves, but if they want to update the rect at the same time, `popup_unconstrain` is called twice and the popup can move even if it wouldn’t really have to (as far as the user cares).
